cub/cub/device/dispatch/dispatch_merge_sort.cuh

```diff
@@ -43,13 +43,99 @@
 #include <cub/util_math.cuh>
 #include <cub/util_namespace.cuh>
 
-#include <thrust/system/cuda/detail/core/triple_chevron_launch.h>
 #include <thrust/detail/integer_math.h>
+#include <thrust/system/cuda/detail/core/triple_chevron_launch.h>
 
 CUB_NAMESPACE_BEGIN
 
-template <bool UseVShmem,
-          typename ChainedPolicyT,
+namespace detail
+{
+
+/**
+ * @brief Helper class template that provides two agent template instantiations: one instantiated with the default
+ * policy and one with the fallback policy. This helps to avoid having to enlist all the agent's template parameters
+ * twice: once for the default agent and once for the fallback agent
+ */
+template <typename DefaultPolicyT, typename FallbackPolicyT, template <typename...> class AgentT, typename... AgentParamsT>
+struct dual_policy_agent_helper_t
+{
+  using default_agent_t  = AgentT<DefaultPolicyT, AgentParamsT...>;
+  using fallback_agent_t = AgentT<FallbackPolicyT, AgentParamsT...>;
+
+  static constexpr auto default_size  = sizeof(typename default_agent_t::TempStorage);
+  static constexpr auto fallback_size = sizeof(typename fallback_agent_t::TempStorage);
+};
+
+/**
+ * @brief Helper class template for merge sort-specific virtual shared memory handling. The merge sort algorithm in its
+ * current implementation relies on the fact that both the sorting as well as the merging kernels use the same tile
+ * size. This circumstance needs to be respected when determining whether the fallback policy for large user types is
+ * applicable: we must either use the fallback for both or for none of the two agents.
+ */
+template < typename default_policy_t,
+           typename KeyInputIteratorT,
+           typename ValueInputIteratorT,
+           typename KeyIteratorT,
+           typename ValueIteratorT,
+           typename OffsetT,
+           typename CompareOpT,
+           typename KeyT,
+           typename ValueT>
+class merge_sort_vsmem_helper_t
+{
+private:
+  // Default fallback policy with a smaller tile size
+  using fallback_policy_t = cub::detail::policy_wrapper_t<default_policy_t, 64, 1>;
```

Comment:
question: instead of writing these helpers and deviating from the vsmem abstraction API, do you think we might get away with defining an aggregate agent and reusing existing vsmem API?
```cpp

template <typename PolicyT,
          typename KeyInputIteratorT,
          typename ValueInputIteratorT,
          typename KeyIteratorT,
          typename ValueIteratorT,
          typename OffsetT,
          typename CompareOpT,
          typename KeyT,
          typename ValueT>
struct aggregate_agent_t 
{
  using block_sort_agent_t =
    AgentBlockSort<PolicyT,
                   KeyInputIteratorT,
                   ValueInputIteratorT,
                   KeyIteratorT,
                   ValueIteratorT,
                   OffsetT,
                   CompareOpT,
                   KeyT,
                   ValueT>;

  using merge_agent_t = AgentMerge<PolicyT, KeyIteratorT, ValueIteratorT, OffsetT, CompareOpT, KeyT, ValueT>;

  union _TempStorage
  {
    typename block_sort_agent_t::TempStorage block_sort;
    typename merge_agent_t::TempStorage merge;
  };

  /// Alias wrapper allowing storage to be unioned
  struct TempStorage : Uninitialized<_TempStorage> {};
};
```

Reply:
Yeah, I had the same thought in the beginning. Actually, that's what I had started with. But I found it was getting unnecessarily convoluted. I.e., mixing up a "merged agent" for determining the fallback policy but not for the vsmem handling.

I think now we have a better separation of concerns: (1) the logic for determining whether to use a fallback policy or not and (2) the vsmem helper for gmem/smem handling. For (1), we sort of have a special case with merge sort, so we keep it in the realm of the merge sort implementation, for (2) we can use the off-the-shelf vsmem facility.